### PR TITLE
Indexshipper revert locking issues

### PR DESCRIPTION
**Pull request by owen-d**
*Note: Reverting these will reintroduce an older bug (race condition around mmap access), but with less complexity than the current bug. I will followup with a second PR that fixes our underlying issue as an accompanying PR*

This reverts two locking related PRs in the indexshipper. They both attempted to introduce safe locking practices, but the problem is structural and neither fixed the underlying problem. The issue occurs when

```
1) ForEach
  * `RLock`s table1, acquires indexset: tenant="1", the post-compacted user.
    * `Lock`s indexset for tenant="1", the post compacted user
    * Does not yet call `RUnlock`, which will be called at end of query
  * `RUnlock`s table1
2) Sync
  * `RLock`s table1
    * _Blocks_ calling `Lock` on the tenant="1" indexset (still held from (1)
3) ForEach
  * _Blocks_ `RLock`ing table1, intending to acquire the tenant="" indexset for the pre-compacted users
4) Deadlock!

This is because the locking pattern looks like:


* Sync
  * RLock table
    * Lock indexset
    * Unlock indexset
  * RUnlock Table

* ForEach
  * RLock table
  * RUnlock table
    * RLock indexset
    * RUnlock indexset at end of query
```

Note: This can also happen in any order of lock acquisitions on the indexSets because `RWMutex` prioritized write locks:
```
// If a goroutine holds a RWMutex for reading and another goroutine might
// call Lock, no goroutine should expect to be able to acquire a read lock
// until the initial read lock is released. In particular, this prohibits
// recursive read locking. This is to ensure that the lock eventually becomes
// available; a blocked Lock call excludes new readers from acquiring the
// lock.
```

A recent goroutine capture during this deadlock shows:
* Goroutines waiting for read lock on the indexset
```
goroutine profile: total 34599
32434 @ 0x43ccf6 0x44dc7e 0x44dc55 0x46a285 0x1c2af6d 0x1c2af48 0x1c29eff 0x1c2d62e 0x1c4062b 0x1c42723 0x1a430af 0x1c42175 0x1c49bb0 0xbbd99a 0x1c49b05 0x1c30ac7 0xf10570 0x1fbffe2 0xb2a71a 0x1d2bc50 0xb2a71a 0xbe7022 0x1e9f166 0xb2a71a 0xbe762a 0xb2a71a 0xb2e27b 0xb2a71a 0xbe8ac2 0xb2a71a 0xb2a5be 0xae4fdb
#	0x46a284	sync.runtime_SemacquireMutex+0x24									/usr/local/go/src/runtime/sema.go:77
#	0x1c2af6c	sync.(*RWMutex).RLock+0xac										/usr/local/go/src/sync/rwmutex.go:71
#	0x1c2af47	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*table).getOrCreateIndexSet+0x87	/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table.go:251
#	0x1c29efe	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*table).ForEach+0xbe			/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table.go:150
#	0x1c2d62d	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*tableManager).ForEach+0x8d		/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table_manager.go:163
#	0x1c4062a	github.com/grafana/loki/pkg/storage/stores/indexshipper.(*indexShipper).ForEach+0x8a			/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/shipper.go:174
#	0x1c42722	github.com/grafana/loki/pkg/storage/stores/shipper/index.(*querier).QueryPages.func1+0x2e2		/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/shipper/index/querier.go:60
#	0x1a430ae	github.com/grafana/loki/pkg/storage/stores/shipper/util.DoParallelQueries+0x16e				/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/shipper/util/queries.go:39
```

* Goroutine waiting for lock on the underlying indexset
```
1 @ 0x43ccf6 0x44dc7e 0x44dc55 0x46a285 0x488411 0x1c2fe65 0x1c2787d 0x1c26fb5 0x1c26eca 0x1c2acdf 0x1c2df27 0x1c2cfb8 0x46e481
#	0x46a284	sync.runtime_SemacquireMutex+0x24									/usr/local/go/src/runtime/sema.go:77
#	0x488410	sync.(*RWMutex).Lock+0x70										/usr/local/go/src/sync/rwmutex.go:152
#	0x1c2fe64	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*mtxWithReadiness).lock+0x44		/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/util.go:39
#	0x1c2787c	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*indexSet).sync+0x63c		/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/index_set.go:297
#	0x1c26fb4	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*indexSet).syncWithRetry+0xb4	/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/index_set.go:256
#	0x1c26ec9	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*indexSet).Sync+0x29			/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/index_set.go:249
#	0x1c2acde	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*table).Sync+0x2be			/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table.go:237
#	0x1c2df26	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*tableManager).syncTables+0x266	/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table_manager.go:216
#	0x1c2cfb7	github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads.(*tableManager).loop+0x377		/src/enterprise-logs/vendor/github.com/grafana/loki/pkg/storage/stores/indexshipper/downloads/table_manager.go:125

### CODE REVIEW

File: pkg/storage/stores/indexshipper/uploads/index_set.go [hunk 1]
```diff
@@ -65,13 +65,9 @@ func (t *indexSet) Add(idx index.Index) {
 	t.index[idx.Name()] = idx
 }
 
-func (t *indexSet) ForEach(_ context.Context, doneChan <-chan struct{}, callback index.ForEachIndexCallback) error {
+func (t *indexSet) ForEach(callback index.ForEachIndexCallback) error {
```
Comment (trevorwhitney) [inline pkg/storage/stores/indexshipper/uploads/index_set.go:68]: do you know the story behind that unused context? I see we're removing it, but I'm curious why it was ever there in the first place?

File: pkg/storage/stores/indexshipper/uploads/index_set.go [hunk 2]
```diff
@@ -21,7 +21,7 @@ type IndexSet interface {
 	Add(idx index.Index)
 	Upload(ctx context.Context) error
 	Cleanup(indexRetainPeriod time.Duration) error
-	ForEach(ctx context.Context, doneChan <-chan struct{}, callback index.ForEachIndexCallback) error
+	ForEach(callback index.ForEachIndexCallback) error
```
Comment (trevorwhitney) [inline pkg/storage/stores/indexshipper/uploads/index_set.go:24]: is it weird that the uploads and downloads packages have different signatures for this function (one with a context, the other without)? It seems like in the latter that context is only necessary for a logger.

**Review by trevorwhitney — COMMENTED** [review]
hey sorry, didn't have time to finish a full review, I'll circle back. Looks good though!

**Review by trevorwhitney — APPROVED** [review]
LGTM, thanks Owen!

**Comment by owen-d** [comment]
Closing in favor of https://github.com/grafana/loki/pull/7497

**Pull request closed without merging**